core/client/src/main/java/alluxio/client/netty/NettyUnderFileSystemFileReader.java

```diff
@@ -0,0 +1,133 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.netty;
+
+import alluxio.Constants;
+import alluxio.exception.ExceptionMessage;
+import alluxio.network.protocol.RPCErrorResponse;
+import alluxio.network.protocol.RPCFileReadRequest;
+import alluxio.network.protocol.RPCFileReadResponse;
+import alluxio.network.protocol.RPCMessage;
+import alluxio.network.protocol.RPCResponse;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Reader for an under file system file through a worker's data server. This class does not hold
+ * any state specific to the file being read and can be used to read multiple files, but not
+ * concurrently. Close should be called on the reader if the caller no longer intends to use the
+ * reader in order to clean up any lingering data from the last read.
+ */
+@NotThreadSafe
+public final class NettyUnderFileSystemFileReader implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** Netty bootstrap for the connection. */
+  private final Bootstrap mClientBootstrap;
+  /** Handler for Netty messages. */
+  private final ClientHandler mHandler;
+  /** A reference to read response so we can explicitly release the resource after reading. */
+  private RPCFileReadResponse mReadResponse = null;
+
+  /**
+   * Creates a new reader for a file in an under file system through a worker's data server.
+   */
+  public NettyUnderFileSystemFileReader() {
+    mHandler = new ClientHandler();
+    mClientBootstrap = NettyClient.createClientBootstrap(mHandler);
+  }
+
+  /**
+   * Reads data from the specified worker for a file in the under file system.
+   *
+   * @param address the worker address to read from
+   * @param ufsFileId the worker specific file id referencing the file to read
+   * @param offset the offset in the file to read from
+   * @param length the length to read
+   * @return a byte buffer with the requested data, null if EOF is reached
+   * @throws IOException if an error occurs communicating with the worker
+   */
+  public ByteBuffer read(InetSocketAddress address, long ufsFileId, long offset, long length)
+      throws IOException {
+    try {
+      ChannelFuture f = mClientBootstrap.connect(address).sync();
+
+      LOG.debug("Connected to remote machine {}", address);
+      Channel channel = f.channel();
+      SingleResponseListener listener = new SingleResponseListener();
+      mHandler.addListener(listener);
+      channel.writeAndFlush(new RPCFileReadRequest(ufsFileId, offset, length));
```

Comment:
As discussed, given a channel with latency = L seconds, throughput = T bytes/second, frame size = F bytes. It takes X seconds to process F bytes in client.
The maximum throughput we can achieve is: F / (L + L + F/T + X) bytes / second.  

For example, L = 10ms, F= 10MB, T = 1GB/second, X = 1ms, throughput = 10M / (0.02 + 0.01 + 0.001) = 0.3226GB/second. 

Instead, I think we should stream the data from the server to client and buffer the data. We can investigate GRPC streaming or implement something by ourselves in netty